63.CUDAInterop/main.cpp

```diff
@@ -0,0 +1,411 @@
+// Copyright (C) 2018-2020 - DevSH Graphics Programming Sp. z O.O.
+// This file is part of the "Nabla Engine".
+// For conditions of distribution and use, see copyright notice in nabla.h
+
+#define _NBL_STATIC_LIB_
+#include <nabla.h>
+
+#include "nbl/video/CCUDAHandler.h"
+#include "nbl/video/CCUDASharedMemory.h"
+#include "nbl/video/CCUDASharedSemaphore.h"
+
+#include "../common./MonoSystemMonoLoggerApplication.hpp"
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace video;
+
+/*
+The start of the main function starts like in most other example. We ask the
+user for the desired renderer and start it up.
+*/
+
+#define ASSERT_SUCCESS(expr) \
+if (auto re = expr; CUDA_SUCCESS != re) { \
+	const char* name = 0, *str = 0; \
+	cu.pcuGetErrorName(re, &name); \
+	cu.pcuGetErrorString(re, &str); \
+	printf("%s:%d %s:\n\t%s\n", __FILE__, __LINE__, name, str); \
+	abort(); \
+}
+
+#define ASSERT_SUCCESS_NV(expr) \
+if (auto re = expr; NVRTC_SUCCESS != re) { \
+	const char* str = cudaHandler->getNVRTCFunctionTable().pnvrtcGetErrorString(re); \
+	printf("%s:%d %s\n", __FILE__, __LINE__, str); \
+	abort(); \
+}
+
+constexpr uint32_t gridDim[3] = { 4096,1,1 };
+constexpr uint32_t blockDim[3] = { 1024,1,1 };
+size_t numElements = gridDim[0] * blockDim[0];
+size_t size = sizeof(float) * numElements;
+
+#ifndef _NBL_COMPILE_WITH_CUDA_
+static_assert(false);
+#endif
+
+class CUDA2VKApp : public examples::MonoSystemMonoLoggerApplication
+{
+	using base_t = examples::MonoSystemMonoLoggerApplication;
+public:
+	// Generally speaking because certain platforms delay initialization from main object construction you should just forward and not do anything in the ctor
+	using base_t::base_t;
+
+	smart_refctd_ptr<CCUDAHandler> cudaHandler;
+	smart_refctd_ptr<CCUDADevice> cudaDevice;
+	// IUtilities* util;
+	smart_refctd_ptr<ILogicalDevice> logicalDevice;
+	IQueue* queue;
+
+	std::array<smart_refctd_ptr<ICPUBuffer>, 2> cpubuffers;
+	std::array<smart_refctd_ptr<CCUDASharedMemory>, 3> mem = {};
+	smart_refctd_ptr<CCUDASharedSemaphore> cusema;
+
+	smart_refctd_ptr<IGPUBuffer> importedbuf, stagingbuf, stagingbuf2;
+	smart_refctd_ptr<IGPUImage> importedimg;
+	smart_refctd_ptr<ISemaphore> sema;
+	smart_refctd_ptr<IGPUCommandPool> commandPool;
+	smart_refctd_ptr<IGPUCommandBuffer> cmd;
```

Comment:
I'd appreciate a block comment about what will happen in the example and what is what